### ISSUE TITLE
NET-1983 Fix S3966 FN: Support `Dispose` call on underlying disposable when wrapped by disposable

### ISSUE THREAD

**Issue by andrei-epure-sonarsource** [SonarSource/sonar-dotnet#2615]
### Description

[RSPEC-3966 Objects should not be disposed more than once](https://rules.sonarsource.com/csharp/RSPEC-3966)

If `Dispose` is called on the `StreamReader` and then on the `FileStream`, we should trigger (because the `StreamReader` will dispose it's underlying `FileStream`). 

### Repro steps

```
  FileStream su = File.Open("c:\temp2.txt", FileMode.Open, FileAccess.Read, FileShare.Write);
  StreamReader r = new StreamReader(su);
  r.Dispose();
  su.Dispose(); // Should raise, but FN
```

### Expected behavior

This is the same as having
```
using (var stream = File.Open("c:\temp2.txt", FileMode.Open, FileAccess.Read, FileShare.Write))
{
  using (var reader = new StreamReader(stream))
  {
  }
}
```
so it should raise.
### Actual behavior

Does not raise.

### Known workarounds

Use `using` pattern.

### Related information

* SonarC# Version 7.16

Came from support via @eric-legoff-sonarsource . See [reproducer on sonarcloud](https://sonarcloud.io/code?id=S3966&selected=S3966%3AProgram.cs)

**Comment by pavel-mikula-sonarsource** [comment]
As a side effect of [this community post](https://community.sonarsource.com/t/false-positive-with-c-using-refactor-this-code-to-make-sure-stream-is-disposed-only-once-csharpsquid-s3966/28164), the issue is also not raised as False-Negative when `return` statement is present inside.

```C#
using (var stream = File.Open("c:\temp2.txt", FileMode.Open, FileAccess.Read, FileShare.Write))
{
    using (var reader = new StreamReader(stream)) // FN due to return. It works when return is removed.
    {
        return; 
    }
}
```
and similar situation with C# 8 syntax.
``` C#
internal static async Task<string> ReadDataFromFileAsync(string path)
{
    await using var stream = Assembly.GetExecutingAssembly().GetManifestResourceStream($"ThePath.{path}");
    using (var reader = new StreamReader(stream))
        return await reader.ReadToEndAsync().ConfigureAwait(false); // FN due to return statement
}
```

When the `return` is outside the `using` statement, it works as expected
``` C#
internal static async Task<string> ReadDataFromFileAsync(string path)
{
    await using var stream = Assembly.GetExecutingAssembly().GetManifestResourceStream($"ThePath.{path}");
    string ret;
    using (var reader = new StreamReader(stream)) // Issue is raised here
        ret = await reader.ReadToEndAsync().ConfigureAwait(false); 
    return ret;
}
```

**Comment by bottle2** [comment]
also consider the constructor with leaveOpen = true, which leaves the stream open (and thus not disposed) after disposing the stream

https://learn.microsoft.com/en-us/dotnet/api/system.io.streamreader.-ctor?view=net-8.0#system-io-streamreader-ctor(system-io-stream-system-text-encoding-system-boolean-system-int32-system-boolean)

**Comment by andrei-epure-sonarsource** [comment]
@pavel-mikula-sonarsource is this something that could be fixed easily with the new engine?

**Comment by Tim-Pohlmann** [comment]
@andrei-epure-sonarsource  We could track if streams are wrapping other streams in a field and check if the wrapping/wrapped symbol has the `Disposed` constraint. It seems not too hard to do.

**Comment by sebastien-marichal** [comment]
Should support types with the `leaveOpen` parameter such as:
- [StreamWriter](https://learn.microsoft.com/en-us/dotnet/api/system.io.streamwriter.-ctor?view=net-8.0#system-io-streamwriter-ctor(system-io-stream-system-text-encoding-system-int32-system-boolean))
- [StreamReader](https://learn.microsoft.com/en-us/dotnet/api/system.io.streamreader.-ctor?view=net-8.0#system-io-streamreader-ctor(system-io-stream-system-text-encoding-system-boolean-system-int32-system-boolean))
- [ZipArchive](https://learn.microsoft.com/en-us/dotnet/api/system.io.compression.ziparchive.-ctor?view=net-8.0#system-io-compression-ziparchive-ctor(system-io-stream-system-io-compression-ziparchivemode-system-boolean))
- [GZipStream](https://learn.microsoft.com/en-us/dotnet/api/system.io.compression.gzipstream.-ctor?view=net-8.0#system-io-compression-gzipstream-ctor(system-io-stream-system-io-compression-compressionlevel-system-boolean))
- [CryptoStream](https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.cryptostream.-ctor?view=net-8.0#system-security-cryptography-cryptostream-ctor(system-io-stream-system-security-cryptography-icryptotransform-system-security-cryptography-cryptostreammode-system-boolean))

And maybe others. Be carefully the default value for `leaveOpen` may not be the same for all.

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1983](https://sonarsource.atlassian.net/browse/NET-1983)

[NET-1983]: https://sonarsource.atlassian.net/browse/NET-1983?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by pavel-mikula-sonarsource** [comment]
Moved
